src/main/xlat_func.c

```diff
@@ -841,20 +841,57 @@ static xlat_action_t xlat_randstr(TALLOC_CTX *ctx, fr_cursor_t *out,
  *
  * Example: "%{urlquote:http://example.org/}" == "http%3A%47%47example.org%47"
  */
-static ssize_t urlquote_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-			     UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-			     UNUSED REQUEST *request, char const *fmt)
+static xlat_action_t xlat_urlquote(TALLOC_CTX *ctx, fr_cursor_t *out,
+				   REQUEST *request, UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+				   fr_value_box_t **in)
 {
-	char const 	*p;
-	char		*out_p = *out;
-	size_t	freespace = outlen;
+	char const 	*p, *end;
+	char		*buff, *buff_p;
+	size_t		outlen = 0;
+	fr_value_box_t	*vb;
 
-	if (outlen <= 1) return 0;
+	/*
+	 * Nothing to do if input is empty
+	 */
+	if (!(*in)) {
+		return XLAT_ACTION_DONE;
+	}
 
-	p = fmt;
-	while (*p && (--freespace > 0)) {
+	/*
+	 * Concatenate all input
+	 */
+	if (fr_value_box_list_concat(ctx, *in, in, FR_TYPE_STRING, true) < 0) {
+		RPEDEBUG("Failed concatenating input");
+		return XLAT_ACTION_FAIL;
+	}
+
+	p = (*in)->vb_strvalue;
+	end = p + (*in)->vb_length;
+
+	/*
+	 * Calculate size of output
+	 */
+	while (p < end) {
+		if (isalnum(*p) ||
+		    *p == '-' ||
+		    *p == '_' ||
+		    *p == '.' ||
+		    *p == '~') {
+			outlen++;
+		} else {
+			outlen += 3;
+		}
+		p++;
+	}
+
+	buff = buff_p = talloc_zero_size(NULL, sizeof(char) * (outlen + 1));
```

Comment:
you should really use talloc_array here specifying a char type (likewise with the other xlats), just so that the type of the buffer is set to char.